src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/AvoidPotentiallyExpensiveCallWhenLogging.cs

```diff
@@ -0,0 +1,333 @@
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!RequiredSymbols.TryGetSymbols(context.Compilation, out var symbols))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(AnalyzeInvocation, OperationKind.Invocation);
+
+            void AnalyzeInvocation(OperationAnalysisContext context)
+            {
+                var invocation = (IInvocationOperation)context.Operation;
+
+                // Check if invocation is a logging invocation and capture the log level (either as IOperation or as int, depending if it is dynamic or not).
+                // Use these to check if the logging invocation is guarded by 'ILogger.IsEnabled' and bail out if it is.
+                if (!symbols.TryGetLogLevel(invocation, out var logLevelArgumentOperation, out var logLevel) ||
+                    symbols.IsGuardedByIsEnabled(currentOperation: invocation, logInvocation: invocation, logLevel, logLevelArgumentOperation))
+                {
+                    return;
+                }
+
+                var arguments = invocation.Arguments.Skip(invocation.IsExtensionMethodAndHasNoInstance() ? 1 : 0);
+
+                // At this stage we have a logging invocation that is not guarded.
+                // Check each argument if it is potentially expensive.
+                foreach (var argument in arguments)
+                {
+                    // Check the argument value after conversions to prevent noise (e.g. implicit conversions from null or from int to EventId).
+                    if (IsPotentiallyExpensive(argument.Value.WalkDownConversion()))
+                    {
+                        // Filter out implicit operations in the case of params arguments.
+                        // If we would report the diagnostic on the implicit argument operation, it would flag the whole invocation.
+                        var explicitDescendants = argument.Value.GetTopmostExplicitDescendants();
+
+                        if (!explicitDescendants.IsEmpty)
+                        {
+                            context.ReportDiagnostic(explicitDescendants[0].CreateDiagnostic(Rule));
+                        }
+                    }
+                }
+            }
+        }
+
+        private static bool IsPotentiallyExpensive(IOperation? operation)
+        {
+            if (operation is null
+                // Implicit params array creation is treated as not expensive. This would otherwise cause a lot of noise.
+                or IArrayCreationOperation { IsImplicit: true, Initializer.ElementValues.IsEmpty: true }
```

Comment:
Same here, without this, all log calls that take a params array would always be flagged, even if they do not capture anything, e.g.:

```C#
void M(ILogger logger)
{
    logger.LogInformation("literal");
}
```